test/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD100AsyncVoidMethodAnalyzerTests.cs

```diff
@@ -24,6 +24,43 @@ class Test {
             await Verify.VerifyAnalyzerAsync(test, expected);
         }
 
+        [Fact]
+        public async Task ReportWarningOnAsyncVoidLocalFunction()
+        {
+            var test = @"
+using System;
+
+class Test {
+    void M() {
+        F();
+
+        async void F() {}
+    }
+}
+";
+            CodeAnalysis.Testing.DiagnosticResult expected = Verify.Diagnostic().WithLocation(8, 20);
+            await Verify.VerifyAnalyzerAsync(test, expected);
+        }
+
+        [Fact]
+        public async Task ReportWarningOnAsyncVoidLambda()
+        {
+            var test = @"
+using System;
+using System.Threading.Tasks;
+
+class Test {
+    void M() {
+        F(async void () => await Task.Delay(0));
```

Comment:
This is invalid syntax, and the cause for the test failure in CI. You mentioned you couldn't repro it locally, but that really surprises me, since it isn't valid in any version of the language or compiler AFAIK. [Even Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gVgG4aayGAVGXDAwDeNBmKYoGAWQAUASmGjxygGIziADgkN5DALwA+JgE4mSAHQARGABtsATxl05cjtSXKxxSWuakG2ApCAL4eDKHUEUA==) shows this failing.

If you just remove the `void` modifier (which lambdas aren't allowed to specify), it works.

Reply:
In C# 10 it's possible to specify the return type of a lambda in code:

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions#explicit-return-type

That Sharplab code works if you set the compiler to main and the output to something other than ASM:

https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAlAVgG5NNlcAVAUwGcAXXAb01zfwBZcBZACgEpmrdiIBivBAA5OuAbgC8APnwBOfADYAdABFqAGwgBPXtn79SGYSLYIu4gklwRBTAL5Xc7jF6A=

I think the problem in this case is the version of the compiler being used differs in CI from what I have locally (int.main).

Removing the explicit return type is fine however, as it's optional in that case.

Reply:
Very strange, and disappointing that the tests aren't effectively pinning whatever versions vary between our official builds and your local build. We are pinning the compiler. I would have thought the langversion was effectively pinned as well.